lib/multiplexer/proxyline.go

```diff
@@ -33,27 +35,50 @@ import (
 	"github.com/gravitational/trace"
 )
 
+// PP2Type is the PROXY protocol v2 TLV type
+type PP2Type byte
+
 const (
 	// TCP4 is TCP over IPv4
 	TCP4 = "TCP4"
 	// TCP6 is tCP over IPv6
 	TCP6 = "TCP6"
 	// Unknown is unsupported or unknown protocol
 	UNKNOWN = "UNKNOWN"
+
+	PP2TypeNOOP PP2Type = 0x04 // No-op used for padding
+
+	// Known custom types, spec allows to use 0xE0 - 0xEF for custom types
+	PP2TypeGCP   PP2Type = 0xE0 // https://cloud.google.com/vpc/docs/configure-private-service-connect-producer
+	PP2TypeAWS   PP2Type = 0xEA // https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-target-groups.html
+	PP2TypeAzure PP2Type = 0xEE // https://learn.microsoft.com/en-us/azure/private-link/private-link-service-overview
+
+	PP2TypeTeleport PP2Type = 0xE4 // Teleport own type for transferring our custom data
 )
 
 var (
 	proxyCRLF = "\r\n"
 	proxySep  = " "
+
+	// ErrTruncatedTLV is returned when there's no enough bytes to read full TLV
+	ErrTruncatedTLV = errors.New("TLV value was truncated")
 )
 
-// ProxyLine is HA Proxy protocol version 1
-// https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt
+// ProxyLine implements PROXY protocol version 1 and 2
+// Spec: https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt
 // Original implementation here: https://github.com/racker/go-proxy-protocol
+// TLV: https://github.com/pires/go-proxyproto
 type ProxyLine struct {
 	Protocol    string
 	Source      net.TCPAddr
 	Destination net.TCPAddr
+	TLVs        []TLV
+}
+
+// TLV (Type-Length-Value) is an extension mechanism in PROXY protocol v2, see end of section 2.2
+type TLV struct {
```

Comment:
It's not strictly `TLV` in our code as we don't store length explicitly here anymore, right? Maybe `Extension` would be better name?

Reply:
We still store length in the byte slice, so this struct defines type, value and length of the value :) 
I really don't want to blur terminology here and prefer to stick to spec naming